packages/blade/src/components/Typography/Heading/Heading.tsx

```diff
@@ -0,0 +1,85 @@
+import type { ReactElement } from 'react';
+import type { TextTypes } from '../../../tokens/theme/theme';
+import getPlatformType from '../../../utils/getPlatformType';
+import type { Theme } from '../../BladeProvider';
+import BaseText from '../BaseText';
+import type { BaseTextProps } from '../BaseText/BaseText';
+
+export type HeadingCommonProps = {
+  type: TextTypes;
+  children?: React.ReactNode;
+};
+
+type HeadingNormalVariant = HeadingCommonProps & {
+  variant: 'small' | 'medium' | 'large';
+  weight: keyof Theme['typography']['fonts']['weight'];
+};
+
+type HeadingSubHeadingVariant = HeadingCommonProps & {
+  variant: 'subheading';
+  weight: keyof Pick<Theme['typography']['fonts']['weight'], 'bold'>;
+};
+
+/**
+ * Conditionally changing props based on variant.
+ * Overloads or union gives wrong intellisense.
+ */
+export type HeadingProps<T> = T extends {
+  variant: infer Variant;
+}
+  ? Variant extends 'small' | 'medium' | 'large'
+    ? HeadingNormalVariant
+    : Variant extends 'subheading'
+    ? HeadingSubHeadingVariant
+    : T
+  : T;
+
+const getProps = <T extends { variant: 'small' | 'medium' | 'large' | 'subheading' }>({
```

Comment:
type HeadingVariants = 'small' | 'medium' | 'large' | 'subheading'; 

```
<T extends { variant: HeadingVariants }>
```

Reply:
Since both `subheading` and `small` are `h6`, `subheading` being the more diminutive element with the lesser fontSize and lineHeight, I feel it would make more sense to start with `subHeading` and go in an ascending order of hierarchy and end with `large`.

`<T extends { variant: 'subheading' | 'small' | 'medium' | 'large' }>`

Reply:
arre sorting is fine, I'm more concerned about the variant `subheading` itself. like `small`, `medium`, `large` and `subheading` aren't going well together

Reply:
Oh I see, you are saying that `subheading` doesn't sound like a variant? 

But to be honest, subheading is the only one which sounds like a variant 😆 `small` `medium` `large` is more like sizes for the title. 

Although agree that the naming aren't going together intuitively.

Reply:
I know chakra does it with `xs` to `xl` but they call their prop `size` instead of `variant`.
https://chakra-ui.com/docs/components/typography/heading

Reply:
added in b209162117341eb6837234bfe411256e515f692f